editor/plugins/editor_script_plugin.cpp

```diff
@@ -0,0 +1,70 @@
+/*                         This file is part of:                          */
+/*                             GODOT ENGINE                               */
+/*                        https://godotengine.org                         */
+/**************************************************************************/
+/* Copyright (c) 2014-present Godot Engine contributors (see AUTHORS.md). */
+/* Copyright (c) 2007-2014 Juan Linietsky, Ariel Manzur.                  */
+/*                                                                        */
+/* Permission is hereby granted, free of charge, to any person obtaining  */
+/* a copy of this software and associated documentation files (the        */
+/* "Software"), to deal in the Software without restriction, including    */
+/* without limitation the rights to use, copy, modify, merge, publish,    */
+/* distribute, sublicense, and/or sell copies of the Software, and to     */
+/* permit persons to whom the Software is furnished to do so, subject to  */
+/* the following conditions:                                              */
+/*                                                                        */
+/* The above copyright notice and this permission notice shall be         */
+/* included in all copies or substantial portions of the Software.        */
+/*                                                                        */
+/* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,        */
+/* EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF     */
+/* MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. */
+/* IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY   */
+/* CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,   */
+/* TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE      */
+/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                 */
+/**************************************************************************/
+
+#include "editor_script_plugin.h"
+
+#include "core/io/resource_loader.h"
+#include "editor/editor_command_palette.h"
+#include "editor/editor_interface.h"
+#include "editor/editor_script.h"
+
+Ref<EditorScript> create_instance(const StringName &p_name) {
+	Ref<EditorScript> es;
+	es.instantiate();
+	Error err;
+	Ref<Script> scr = ResourceLoader::load(ScriptServer::get_global_class_path(p_name), "Script", ResourceFormatLoader::CACHE_MODE_REUSE, &err);
+	if (err == OK) {
+		es->set_script(scr);
+	}
+	return es;
+}
+
+void EditorScriptPlugin::run_command(const StringName &p_name) {
+	create_instance(p_name)->run();
+}
+
+void EditorScriptPlugin::command_palette_about_to_popup() {
+	for (const StringName &command : commands) {
+		EditorInterface::get_singleton()->get_command_palette()->remove_command("editor_scripts/" + command);
+	}
+	commands.clear();
+	ScriptServer::get_inheriters_list(SNAME("EditorScript"), &commands);
```

Comment:
It would be nice to avoid doing this *every time* the palette pops up, especially when you are instantiating every script (which can be heavy due to preloads).

Reply:
I totally agree, it would be nice to avoid that. But I didn't really find any signals that could be used to invalidate a cache. All the stuff on `EditorFileSystem` seems to trigger so often, that the cache would basically never be used. Open for ideas here.

Reply:
Maybe @Hilderin has ideas after poking quite a bit at `EditorFileSystem` logic?

Reply:
Could cut `_get_name()` and just format the class name with spaces for the display name, then there's no need to instantiate scripts before a command is ran. The class name is required anyway and the command palette api is still available through plugins if display name customization is important

Reply:
That feels too limiting, looking at how I would use the feature I would adopt a common naming scheme for the class names e.g. prefixing them with `Command`. I don't want that noise in the command palette.

Also just looking at one example that I actually used at one point: `CommandRecalculate3DNavigationMeshForIsometricTilemapInOpenScene` isn't a class name I want to have in my project.

Reply:
Maybe we could just keep a cache of previously ran EditorScripts. If you run a script once, it gets registered in some cache file and editor will automatically add it to command palette. It's cheap, the only downside is that you need to run the script normally once.

Reply:
> the only downside is that you need to run the script normally once

There is no "normal" way to run an editor script when you use an external editor. Having to switch to the built-in and run all of my 40 editor scripts before I can use them each time I make a new godot project would make this feature as good as non-existent for me. Has to be a better way 🙂

Reply:
That honestly sounds very confusing UX wise. If that's the alternative I'd rather remove the `_get_name` feature.

Reply:
Yeah, it would be the same as not showing any scenes in the Ctrl+Shift+A dialog before the user instantiates them through the filesystem dock